File: 1871-jump-game-vii/1871-jump-game-vii.py
class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:
        n, cnt = len(s), 0
        dp = [False] * n
        dp[0] = True
        for i in range(1, n):
            if i >= minJump and i - minJump >= 0 and dp[i - minJump] == True:
                cnt += 1
            if i > maxJump and i - maxJump - 1 >= 0 and dp[i - maxJump - 1] == True:
                cnt -= 1
            if cnt > 0 and s[i] == '0':
                dp[i] = True
        return dp[-1]

File: 1871-jump-game-vii/1871-jump-game-vii.kt
class Solution {
    fun canReach(s: String, minJump: Int, maxJump: Int): Boolean {
        val n = s.length
        var windowOfJumps = 0
        val dp = BooleanArray(n){false}
        dp[0] = true
        for (jumpSpot in 1..n-1) {
            if (jumpSpot >= minJump) {
                windowOfJumps += if (jumpSpot - minJump >= 0 && dp[jumpSpot - minJump] === true) 1 else 0
            }
            if (jumpSpot > maxJump) {
                windowOfJumps -= if (jumpSpot - maxJump - 1 >= 0 && dp[jumpSpot - maxJump - 1] === true) 1 else 0
            }
            if (windowOfJumps > 0 && s[jumpSpot] === '0'.toChar()) {
                dp[jumpSpot] = true
            }
        }
        return dp[n-1]
    }
}